File: Problemset/relative-sort-array/relative-sort-array.go

// @Title: 数组的相对排序 (Relative Sort Array)
// @Author: KivenC
// @Date: 2019-07-18 15:10:00
// @Runtime: 8 ms
// @Memory: 6.1 MB

func relativeSortArray(arr1 []int, arr2 []int) []int {
    // 统计 arr1 中元素
    counter := map[int]int{}
    for _, k := range arr1 {
        counter[k]++
    }
    res := []int{}
    // 先将 arr2 中存在的元素排好
    for _, k := range arr2 {
        for counter[k] > 0 {
            res = append(res, k)
            counter[k]--
        }
    }
    // 将 arr2 中不存在的元素按升序排好
    for i := 0; i <= 1000; i++ {
        for counter[i] > 0 {
            res = append(res, i)
            counter[i]--
        }
    }
    return res
}

File: Problemset/relative-sort-array/relative-sort-array.py

# @Title: 数组的相对排序 (Relative Sort Array)
# @Author: KivenC
# @Date: 2019-07-18 14:56:57
# @Runtime: 48 ms
# @Memory: 13.1 MB

class Solution:
    def relativeSortArray(self, arr1: List[int], arr2: List[int]) -> List[int]:
        # 统计 arr1 中的元素，按照 arr2 相对顺序填放，剩下的元素按照升序放置
        import collections
        counter = collections.Counter(arr1)
        res = []
        for a in arr2:
            res.extend([a] * counter[a])
            del counter[a]
        for k in sorted(counter.keys()):
            res.extend([k] * counter[k])
        return res
